oauth2/oauth2-core/src/main/java/org/springframework/security/oauth2/core/endpoint/MapOAuth2AccessTokenResponseConverter.java

```diff
@@ -38,59 +38,112 @@
 public final class MapOAuth2AccessTokenResponseConverter
 		implements Converter<Map<String, String>, OAuth2AccessTokenResponse> {
 
+	private final ObjectValuesConverter objectValuesConverter = new ObjectValuesConverter();
+
 	private static final Set<String> TOKEN_RESPONSE_PARAMETER_NAMES = new HashSet<>(
 			Arrays.asList(OAuth2ParameterNames.ACCESS_TOKEN, OAuth2ParameterNames.EXPIRES_IN,
 					OAuth2ParameterNames.REFRESH_TOKEN, OAuth2ParameterNames.SCOPE, OAuth2ParameterNames.TOKEN_TYPE));
 
+	/**
+	 * @deprecated Use {@link MapOAuth2AccessTokenResponseConverter#of()} instead.
+	 */
+	@Deprecated
+	public MapOAuth2AccessTokenResponseConverter() {
+		// Intentionally left blank.
+	}
+
 	@Override
 	public OAuth2AccessTokenResponse convert(Map<String, String> tokenResponseParameters) {
-		String accessToken = tokenResponseParameters.get(OAuth2ParameterNames.ACCESS_TOKEN);
-		OAuth2AccessToken.TokenType accessTokenType = getAccessTokenType(tokenResponseParameters);
-		long expiresIn = getExpiresIn(tokenResponseParameters);
-		Set<String> scopes = getScopes(tokenResponseParameters);
-		String refreshToken = tokenResponseParameters.get(OAuth2ParameterNames.REFRESH_TOKEN);
-		Map<String, Object> additionalParameters = new LinkedHashMap<>();
-		for (Map.Entry<String, String> entry : tokenResponseParameters.entrySet()) {
-			if (!TOKEN_RESPONSE_PARAMETER_NAMES.contains(entry.getKey())) {
-				additionalParameters.put(entry.getKey(), entry.getValue());
-			}
-		}
-		// @formatter:off
-		return OAuth2AccessTokenResponse.withToken(accessToken)
-				.tokenType(accessTokenType)
-				.expiresIn(expiresIn)
-				.scopes(scopes)
-				.refreshToken(refreshToken)
-				.additionalParameters(additionalParameters)
-				.build();
-		// @formatter:on
+		return this.objectValuesConverter.convert(tokenResponseParameters);
 	}
 
-	private OAuth2AccessToken.TokenType getAccessTokenType(Map<String, String> tokenResponseParameters) {
-		if (OAuth2AccessToken.TokenType.BEARER.getValue()
-				.equalsIgnoreCase(tokenResponseParameters.get(OAuth2ParameterNames.TOKEN_TYPE))) {
-			return OAuth2AccessToken.TokenType.BEARER;
-		}
-		return null;
+	/**
+	 * Create a new {@link Converter} instance used to convert to an
+	 * {@link OAuth2AccessTokenResponse} that supports both scalar values and nested
+	 * objects.
+	 * @return A new {@link Converter}
+	 */
+	public static Converter<Map<String, ?>, OAuth2AccessTokenResponse> of() {
```

Comment:
Providing a factory method instead of deprecating the class will keep this class around for 6.0 which is not what we want given that the generic type parameters in the class are incorrect. We want this class to go away in 6.0. So let's deprecate the class and introduce a new one with the correct generic type parameters.